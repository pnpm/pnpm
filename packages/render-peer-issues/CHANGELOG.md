# @pnpm/render-peer-issues

## 4.0.4

### Patch Changes

- Updated dependencies [43ce9e4a6]
  - @pnpm/types@9.4.0

## 4.0.3

### Patch Changes

- Updated dependencies [d774a3196]
  - @pnpm/types@9.3.0

## 4.0.2

### Patch Changes

- Updated dependencies [aa2ae8fe2]
  - @pnpm/types@9.2.0

## 4.0.1

### Patch Changes

- Updated dependencies [a9e0b7cbf]
  - @pnpm/types@9.1.0

## 4.0.0

### Major Changes

- eceaa8b8b: Node.js 14 support dropped.

### Patch Changes

- Updated dependencies [eceaa8b8b]
  - @pnpm/types@9.0.0

## 3.0.3

### Patch Changes

- Updated dependencies [b77651d14]
  - @pnpm/types@8.10.0

## 3.0.2

### Patch Changes

- Updated dependencies [702e847c1]
  - @pnpm/types@8.9.0

## 3.0.1

### Patch Changes

- Updated dependencies [844e82f3a]
  - @pnpm/types@8.8.0

## 3.0.0

### Major Changes

- f884689e0: Require `@pnpm/logger` v5.

## 2.1.2

### Patch Changes

- Updated dependencies [d665f3ff7]
  - @pnpm/types@8.7.0

## 2.1.1

### Patch Changes

- Updated dependencies [156cc1ef6]
  - @pnpm/types@8.6.0

## 2.1.0

### Minor Changes

- c990a409f: Print the versions of packages in peer dependency warnings and errors.

## 2.0.6

### Patch Changes

- Updated dependencies [c90798461]
  - @pnpm/types@8.5.0

## 2.0.5

### Patch Changes

- Updated dependencies [8e5b77ef6]
  - @pnpm/types@8.4.0

## 2.0.4

### Patch Changes

- Updated dependencies [2a34b21ce]
  - @pnpm/types@8.3.0

## 2.0.3

### Patch Changes

- Updated dependencies [fb5bbfd7a]
  - @pnpm/types@8.2.0

## 2.0.2

### Patch Changes

- Updated dependencies [4d39e4a0c]
  - @pnpm/types@8.1.0

## 2.0.1

### Patch Changes

- Updated dependencies [18ba5e2c0]
  - @pnpm/types@8.0.1

## 2.0.0

### Major Changes

- 542014839: Node.js 12 is not supported.

### Patch Changes

- Updated dependencies [d504dc380]
- Updated dependencies [542014839]
  - @pnpm/types@8.0.0

## 1.1.2

### Patch Changes

- Updated dependencies [b138d048c]
  - @pnpm/types@7.10.0

## 1.1.1

### Patch Changes

- Updated dependencies [26cd01b88]
  - @pnpm/types@7.9.0

## 1.1.0

### Minor Changes

- b5734a4a7: When reporting unmet peer dependency issues, if the peer dependency is resolved not from a dependency installed by the user, then print the name of the parent package that has the bad peer dependency installed as a dependency.

  ![](https://i.imgur.com/0kjij22.png)

### Patch Changes

- Updated dependencies [b5734a4a7]
  - @pnpm/types@7.8.0

## 1.0.2

### Patch Changes

- 6058f76cd: When printing peer dependency issues, print the "\*" range in double quotes. This will make it easier to copy the package resolutions and put them to the end of a `pnpm add` command for execution.

## 1.0.1

### Patch Changes

- a087f339e: A new line should be between the summary about conflicting peers and non-conflicting ones.
- Updated dependencies [6493e0c93]
  - @pnpm/types@7.7.1

## 1.0.0

### Major Changes

- ba9b2eba1: Initial release.

### Patch Changes

- Updated dependencies [ba9b2eba1]
  - @pnpm/types@7.7.0
