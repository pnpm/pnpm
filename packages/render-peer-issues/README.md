# @pnpm/render-peer-issues

> Visualizes peer dependency issues

## Installation

```
pnpm add @pnpm/render-peer-issues
```

## License

[MIT](LICENSE)
