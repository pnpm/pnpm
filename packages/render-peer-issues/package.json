{
  "name": "@pnpm/render-peer-issues",
  "description": "Visualizes peer dependency issues",
  "version": "4.0.3",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib",
    "!*.map"
  ],
  "keywords": [
    "pnpm8"
  ],
  "license": "MIT",
  "engines": {
    "node": ">=16.14"
  },
  "repository": "https://github.com/pnpm/pnpm/blob/main/packages/render-peer-issues",
  "scripts": {
    "_test": "jest",
    "test": "pnpm run compile && pnpm run _test",
    "lint": "eslint \"src/**/*.ts\" \"test/**/*.ts\"",
    "prepublishOnly": "pnpm run compile",
    "compile": "tsc --build && pnpm run lint --fix"
  },
  "homepage": "https://github.com/pnpm/pnpm/blob/main/packages/render-peer-issues#readme",
  "funding": "https://opencollective.com/pnpm",
  "dependencies": {
    "@pnpm/types": "workspace:*",
    "archy": "^1.0.0",
    "chalk": "^4.1.2",
    "cli-columns": "^4.0.0"
  },
  "devDependencies": {
    "@pnpm/render-peer-issues": "workspace:*",
    "@types/archy": "0.0.33",
    "strip-ansi": "^6.0.1"
  },
  "exports": {
    ".": "./lib/index.js"
  }
}
