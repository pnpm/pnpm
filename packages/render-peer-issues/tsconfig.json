{
  "extends": "@pnpm/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": [
    "src/**/*.ts",
    "../../__typings__/**/*.d.ts"
  ],
  "references": [
    {
      "path": "../../config/matcher"
    },
    {
      "path": "../../config/parse-overrides"
    },
    {
      "path": "../error"
    },
    {
      "path": "../types"
    }
  ],
  "composite": true
}
