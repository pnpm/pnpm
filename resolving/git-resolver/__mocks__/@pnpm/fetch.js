module.exports = jest.createMockFromModule('@pnpm/fetch')

// default implementation
module.exports.fetch.mockImplementation(async (_url, _opts) => {
  return { ok: true }
})
