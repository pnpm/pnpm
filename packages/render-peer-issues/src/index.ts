import { PnpmError } from '@pnpm/error'
import { createMatcher } from '@pnpm/matcher'
import {
  type BadPeerDependencyIssue,
  type PeerDependencyIssuesByProjects,
  type PeerDependencyRules,
} from '@pnpm/types'
import { parseOverrides, type VersionOverride } from '@pnpm/parse-overrides'
import archy from 'archy'
import chalk from 'chalk'
import cliColumns from 'cli-columns'
import semver from 'semver'

export function renderPeerIssues (
  peerDependencyIssuesByProjects: PeerDependencyIssuesByProjects,
  opts?: {
    rules?: PeerDependencyRules
    width?: number
  }
) {
  const ignoreMissingPatterns = [...new Set(opts?.rules?.ignoreMissing ?? [])]
  const ignoreMissingMatcher = createMatcher(ignoreMissingPatterns)
  const allowAnyPatterns = [...new Set(opts?.rules?.allowAny ?? [])]
  const allowAnyMatcher = createMatcher(allowAnyPatterns)
  const { allowedVersionsMatchAll, allowedVersionsByParentPkgName } = parseAllowedVersions(opts?.rules?.allowedVersions ?? {})
  const projects = {} as Record<string, PkgNode>
  for (const [projectId, { bad, missing, conflicts, intersections }] of Object.entries(peerDependencyIssuesByProjects)) {
    projects[projectId] = { dependencies: {}, peerIssues: [] }
    for (const [peerName, issues] of Object.entries(missing)) {
      if (
        !conflicts.includes(peerName) &&
        intersections[peerName] == null ||
        ignoreMissingMatcher(peerName)
      ) {
        continue
      }
      for (const issue of issues) {
        createTree(projects[projectId], issue.parents, `${chalk.red('✕ missing peer')} ${formatNameAndRange(peerName, issue.wantedRange)}`)
      }
    }
    for (const [peerName, issues] of Object.entries(bad)) {
      if (allowAnyMatcher(peerName)) continue
      for (const issue of issues) {
        if (allowedVersionsMatchAll[peerName]?.some((range) => semver.satisfies(issue.foundVersion, range))) continue
        const currentParentPkg = issue.parents.at(-1)
        if (currentParentPkg && allowedVersionsByParentPkgName[peerName]?.[currentParentPkg.name]) {
          const allowedVersionsByParent = allowedVersionsByParentPkgName[peerName][currentParentPkg.name]
            .reduce((acc, { targetPkg, parentPkg, ranges }) => {
              if (!parentPkg.pref || currentParentPkg.version &&
                (isSubRange(parentPkg.pref, currentParentPkg.version) || semver.satisfies(currentParentPkg.version, parentPkg.pref))) {
                acc[targetPkg.name] = ranges
              }
              return acc
            }, {} as Record<string, string[]>)
          if (allowedVersionsByParent[peerName]?.some((range) => semver.satisfies(issue.foundVersion, range))) continue
        }
        createTree(projects[projectId], issue.parents, formatUnmetPeerMessage({
          peerName,
          ...issue,
        }))
      }
    }
  }
  const cliColumnsOptions = {
    newline: '\n  ',
    width: (opts?.width ?? process.stdout.columns) - 2,
  }
  return Object.entries(projects)
    .filter(([, project]) => Object.keys(project.dependencies).length > 0)
    .sort(([projectKey1], [projectKey2]) => projectKey1.localeCompare(projectKey2))
    .map(([projectKey, project]) => {
      const summaries = []
      const { conflicts, intersections } = peerDependencyIssuesByProjects[projectKey]
      if (conflicts.length) {
        summaries.push(
          chalk.red(`✕ Conflicting peer dependencies:\n  ${cliColumns(conflicts, cliColumnsOptions)}`)
        )
      }
      if (Object.keys(intersections).length) {
        summaries.push(
          `Peer dependencies that should be installed:\n  ${cliColumns(Object.entries(intersections).map(([name, version]) => formatNameAndRange(name, version)), cliColumnsOptions)}`
        )
      }
      const title = chalk.white(projectKey)
      let summariesConcatenated = summaries.join('\n')
      if (summariesConcatenated) {
        summariesConcatenated += '\n'
      }
      return `${archy(toArchyData(title, project))}${summariesConcatenated}`
    }).join('\n')
}

function formatUnmetPeerMessage (
  { foundVersion, peerName, wantedRange, resolvedFrom }: BadPeerDependencyIssue & {
    peerName: string
  }
) {
  const nameAndRange = formatNameAndRange(peerName, wantedRange)
  if (resolvedFrom && resolvedFrom.length > 0) {
    return `✕ unmet peer ${nameAndRange}: found ${foundVersion} in ${resolvedFrom[resolvedFrom.length - 1].name}`
  }
  return `${chalk.yellowBright('✕ unmet peer')} ${nameAndRange}: found ${foundVersion}`
}

function formatNameAndRange (name: string, range: string) {
  if (range.includes(' ') || range === '*') {
    return `${name}@"${range}"`
  }
  return `${name}@${range}`
}

interface PkgNode {
  peerIssues: string[]
  dependencies: Record<string, PkgNode>
}

function createTree (pkgNode: PkgNode, pkgs: Array<{ name: string, version: string }>, issueText: string) {
  const [pkg, ...rest] = pkgs
  const label = `${pkg.name} ${chalk.grey(pkg.version)}`
  if (!pkgNode.dependencies[label]) {
    pkgNode.dependencies[label] = { dependencies: {}, peerIssues: [] }
  }
  if (rest.length === 0) {
    pkgNode.dependencies[label].peerIssues.push(issueText)
    return
  }
  createTree(pkgNode.dependencies[label], rest, issueText)
}

function toArchyData (depName: string, pkgNode: PkgNode): archy.Data {
  const result: Required<archy.Data> = {
    label: depName,
    nodes: [],
  }
  for (const wantedPeer of pkgNode.peerIssues) {
    result.nodes.push(wantedPeer)
  }
  for (const [depName, node] of Object.entries(pkgNode.dependencies)) {
    result.nodes.push(toArchyData(depName, node))
  }
  return result
}

type AllowedVersionsByParentPkgName = Record<string, Record<string, Array<Required<Pick<VersionOverride, 'parentPkg' | 'targetPkg'>> & { ranges: string[] }>>>

function parseAllowedVersions (allowedVersions: Record<string, string>) {
  const overrides = tryParseAllowedVersions(allowedVersions)
  const allowedVersionsMatchAll: Record<string, string[]> = {}
  const allowedVersionsByParentPkgName: AllowedVersionsByParentPkgName = {}
  for (const { parentPkg, targetPkg, newPref } of overrides) {
    const ranges = parseVersions(newPref)
    if (!parentPkg) {
      allowedVersionsMatchAll[targetPkg.name] = ranges
      continue
    }
    if (!allowedVersionsByParentPkgName[targetPkg.name]) {
      allowedVersionsByParentPkgName[targetPkg.name] = {}
    }
    if (!allowedVersionsByParentPkgName[targetPkg.name][parentPkg.name]) {
      allowedVersionsByParentPkgName[targetPkg.name][parentPkg.name] = []
    }
    allowedVersionsByParentPkgName[targetPkg.name][parentPkg.name].push({
      parentPkg,
      targetPkg,
      ranges,
    })
  }
  return {
    allowedVersionsMatchAll,
    allowedVersionsByParentPkgName,
  }
}

function tryParseAllowedVersions (allowedVersions: Record<string, string>): VersionOverride[] {
  try {
    return parseOverrides(allowedVersions ?? {})
  } catch (err) {
    throw new PnpmError('INVALID_ALLOWED_VERSION_SELECTOR',
      `${(err as PnpmError).message} in pnpm.peerDependencyRules.allowedVersions`)
  }
}

function parseVersions (versions: string): string[] {
  return versions.split('||').map(v => v.trim())
}

function isSubRange (superRange: string | undefined, subRange: string) {
  return !superRange ||
  subRange === superRange ||
  semver.validRange(subRange) != null &&
  semver.validRange(superRange) != null &&
  semver.subset(subRange, superRange)
}
