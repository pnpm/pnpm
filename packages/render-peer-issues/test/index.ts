import { renderPeerIssues } from '@pnpm/render-peer-issues'
import stripAnsi from 'strip-ansi'

test('renderPeerIssues()', () => {
  expect(stripAnsi(renderPeerIssues({
    'packages/0': {
      conflicts: ['eee'],
      intersections: { ddd: '^1.0.0' },
      bad: {},
      missing: {
        ddd: [
          {
            parents: [
              {
                name: 'zzz',
                version: '1.0.0',
              },
            ],
            optional: false,
            wantedRange: '^1.0.0',
          },
        ],
        eee: [
          {
            parents: [
              {
                name: 'zzz',
                version: '1.0.0',
              },
            ],
            optional: false,
            wantedRange: '^1.0.0',
          },
          {
            parents: [
              {
                name: 'www',
                version: '1.0.0',
              },
            ],
            optional: false,
            wantedRange: '^2.0.0',
          },
        ],
      },
    },
    '.': {
      missing: {
        aaa: [
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
              {
                name: 'yyy',
                version: '1.0.0',
              },
            ],
            optional: false,
            wantedRange: '>=1.0.0 <3.0.0',
          },
        ],
      },
      bad: {
        bbb: [
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
            ],
            foundVersion: '2',
            resolvedFrom: [],
            optional: false,
            wantedRange: '^1.0.0',
          },
        ],
        ccc: [
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
              {
                name: 'yyy',
                version: '1.0.0',
              },
            ],
            foundVersion: '2',
            resolvedFrom: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
            ],
            optional: false,
            wantedRange: '^1.0.0',
          },
        ],
      },
      conflicts: [],
      intersections: { aaa: '^1.0.0' },
    },
  }, { width: 500 }))).toMatchSnapshot()
})

test('renderPeerIssues() ignore missing', () => {
  expect(stripAnsi(renderPeerIssues({
    '.': {
      missing: {
        aaa: [
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
            ],
            optional: false,
            wantedRange: '>=1.0.0 <3.0.0',
          },
        ],
        '@foo/bar': [
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
            ],
            optional: false,
            wantedRange: '>=1.0.0 <3.0.0',
          },
        ],
      },
      bad: {},
      conflicts: [],
      intersections: {
        aaa: '^1.0.0',
        '@foo/bar': '^1.0.0',
      },
    },
  }, {
    rules: {
      ignoreMissing: ['aaa', '@foo/*'],
    },
    width: 500,
  }))).toBe('')
})

test('renderPeerIssues() allow any version', () => {
  expect(stripAnsi(renderPeerIssues({
    '.': {
      missing: {},
      bad: {
        bbb: [
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
            ],
            foundVersion: '2.0.0',
            resolvedFrom: [],
            optional: false,
            wantedRange: '^1.0.0',
          },
        ],
        '@foo/bar': [
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
            ],
            foundVersion: '2.0.0',
            resolvedFrom: [],
            optional: false,
            wantedRange: '^1.0.0',
          },
        ],
      },
      conflicts: [],
      intersections: {},
    },
  }, {
    rules: {
      allowAny: ['bbb', '@foo/*'],
    },
    width: 500,
  }))).toBe('')
})

test('renderPeerIssues() allowed versions', () => {
  expect(stripAnsi(renderPeerIssues({
    '.': {
      missing: {},
      bad: {
        bbb: [
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
            ],
            foundVersion: '2.0.0',
            resolvedFrom: [],
            optional: false,
            wantedRange: '^1.0.0',
          },
        ],
        '@foo/bar': [
          {
            parents: [
              {
                name: 'aaa',
                version: '1.0.0',
              },
            ],
            foundVersion: '2.0.0',
            resolvedFrom: [],
            optional: false,
            wantedRange: '^1.0.0',
          },
          {
            parents: [
              {
                name: 'yyy',
                version: '1.0.0',
              },
              {
                name: 'xxx',
                version: '1.0.0',
              },
            ],
            foundVersion: '2.0.0',
            resolvedFrom: [],
            optional: false,
            wantedRange: '^1.0.0',
          },
          {
            parents: [
              {
                name: 'ccc',
                version: '3.0.0',
              },
            ],
            foundVersion: '3.0.0',
            resolvedFrom: [],
            optional: false,
            wantedRange: '^1.0.0',
          },
          {
            parents: [
              {
                name: 'ccc',
                version: '2.3.6',
              },
            ],
            foundVersion: '4.0.0',
            resolvedFrom: [],
            optional: false,
            wantedRange: '^1.0.0',
          },
        ],
      },
      conflicts: [],
      intersections: {},
    },
  }, {
    rules: {
      allowedVersions: {
        bbb: '2',
        'xxx>@foo/bar': '2',
        'ccc@3>@foo/bar': '3',
        'ccc@>=2.3.5 <3>@foo/bar': '4',
      },
    },
    width: 500,
  }))).toMatchSnapshot()
})

test('renderPeerIssues() optional peer dependencies are printed only if they are in conflict with non-optional peers', () => {
  expect(stripAnsi(renderPeerIssues({
    '.': {
      missing: {
        aaa: [
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
              {
                name: 'yyy',
                version: '1.0.0',
              },
            ],
            optional: true,
            wantedRange: '^1.0.0',
          },
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
              {
                name: 'yyy',
                version: '1.0.0',
              },
            ],
            optional: false,
            wantedRange: '^2.0.0',
          },
        ],
        bbb: [
          {
            parents: [
              {
                name: 'xxx',
                version: '1.0.0',
              },
            ],
            optional: true,
            wantedRange: '^1.0.0',
          },
        ],
      },
      bad: {},
      conflicts: ['aaa'],
      intersections: {},
    },
    empty: {
      missing: {},
      bad: {},
      conflicts: [],
      intersections: {},
    },
  }, { width: 500 }))).toMatchSnapshot()
})

test('renderPeerIssues() format correctly the version ranges with spaces and "*"', () => {
  expect(stripAnsi(renderPeerIssues({
    '.': {
      conflicts: [],
      intersections: { a: '*', b: '1 || 2' },
      bad: {},
      missing: {
        a: [
          {
            parents: [
              {
                name: 'z',
                version: '1.0.0',
              },
            ],
            optional: false,
            wantedRange: '*',
          },
        ],
        b: [
          {
            parents: [
              {
                name: 'z',
                version: '1.0.0',
              },
            ],
            optional: false,
            wantedRange: '1 || 2',
          },
        ],
      },
    },
  }, { width: 500 }))).toMatchSnapshot()
})
