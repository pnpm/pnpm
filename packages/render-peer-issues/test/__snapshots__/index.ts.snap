// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderPeerIssues() 1`] = `
".
└─┬ xxx 1.0.0
  ├── ✕ unmet peer bbb@^1.0.0: found 2
  └─┬ yyy 1.0.0
    ├── ✕ missing peer aaa@">=1.0.0 <3.0.0"
    └── ✕ unmet peer ccc@^1.0.0: found 2 in xxx
Peer dependencies that should be installed:
  aaa@^1.0.0  

packages/0
├─┬ zzz 1.0.0
│ ├── ✕ missing peer ddd@^1.0.0
│ └── ✕ missing peer eee@^1.0.0
└─┬ www 1.0.0
  └── ✕ missing peer eee@^2.0.0
✕ Conflicting peer dependencies:
  eee  
Peer dependencies that should be installed:
  ddd@^1.0.0  
"
`;

exports[`renderPeerIssues() allowed versions 1`] = `
".
└─┬ aaa 1.0.0
  └── ✕ unmet peer @foo/bar@^1.0.0: found 2.0.0
"
`;

exports[`renderPeerIssues() format correctly the version ranges with spaces and "*" 1`] = `
".
└─┬ z 1.0.0
  ├── ✕ missing peer a@"*"
  └── ✕ missing peer b@"1 || 2"
Peer dependencies that should be installed:
  a@"*"       b@"1 || 2"  
"
`;

exports[`renderPeerIssues() optional peer dependencies are printed only if they are in conflict with non-optional peers 1`] = `
".
└─┬ xxx 1.0.0
  └─┬ yyy 1.0.0
    ├── ✕ missing peer aaa@^1.0.0
    └── ✕ missing peer aaa@^2.0.0
✕ Conflicting peer dependencies:
  aaa  
"
`;
